### ISSUE TITLE
Renaming broken cxxctorinit renames class

### ISSUE THREAD

**Issue by sam-mccall** [clangd/clangd#765]
```
struct A {
  int x;
  A() : x(y) {} // rename x or y
}
```

Renaming either x or y actually renames the struct `A`.
The problem is we don't actually have a CXXCtorInitializer node, so the class claims the range.

However the scary thing is that prepareRename claims we're going to be able to rename x/y. It seems happy to have seen an identifier token (which was part of the failed/recovered parse) and assume it's the name of the enclosing node.

We should do some better checking that these match, and make prepareRename & rename fail. Maybe something as simple as checking that the identifier is the current name?

**Comment by sam-mccall** [comment]
A more robust check (that isn't fooled by name collisions) would be to check whether the name-location of the putative-thing-to-be-renamed is actually the selected location.

But this may rule out some legitimate cases maybe?

@hokein @kirillbobyrev any opinions?

**Comment by hokein** [comment]
I think the general guide: rename/refactoring for broken code is best effort, right?  I agree that we should avoid surprising behaviors in clangd when handling the broken code.
 
In this particular case, we could preserve the CtorInitializer (with a recoveryExpr as its initExpr) in clang AST, but this solution is not scale..

>  Maybe something as simple as checking that the identifier is the current name?

This is trivial and should work for most cases, but not for all cases (e.g. name shadowing) etc.

> A more robust check (that isn't fooled by name collisions) would be to check whether the name-location of the putative-thing-to-be-renamed is actually the selected location.

This seems like a good solution, but the implementation seems not trivial to me -- SelectTree returns a DynNode, which is a generic node for various clang AST nodes, we might end up with many ad-hocs to check the rename-triggering location is the name-loc of that DynNode.

Another approach is that we can check whether the rename-triggering range is one of the rename occurrences, I think it is reasonable mental model to assume that if rename is allowed when the user triggers it on a position, that position should be renamed as well.

**Comment by sam-mccall** [comment]
> rename/refactoring for broken code is best effort, right?

Yes, ish... we're definitely within our rights to **reject** operations around broken code.
I think we should spend some effort to avoid applying broken refactorings though, especially given circumstances here (rename is heavily used, broken ctorinit is fairly common, result is bizarre).

> Another approach is that we can check whether the rename-triggering range is one of the rename occurrences, I think it is reasonable mental model to assume that if rename is allowed when the user triggers it on a position, that position should be renamed as well.

Oh, and we always compute such ranges internally even for prepareRename, right? This seems like an excellent solution that seems to work for all cases.

(Both other options you mention also seem much better than today - we could catch many common cases but not all.)

> In this particular case, we could preserve the CtorInitializer (with a recoveryExpr as its initExpr) in clang AST, but this solution is not scale

Agree this doesn't scale to rename-on-all-broken-code, but I think this *is* worthwhile as it will affect all-operations-on-broken-ctorinit. I'll try putting together a patch.
